scripts/evolution.py

```diff
@@ -0,0 +1,441 @@
+                        beginning_of_time,
+                        run_directory,
+                        evolution_data_array,
+                        fast_addressing_of_data_array,
+                        enable_feasible_predictor=False,
+                        evaluation_limit=float("inf"),
+                        black_box_function=None,
+                        number_of_cpus=0):
+    """
+    Evaluate a list of configurations using the black-box function being optimized.
+    This method avoids evaluating repeated points by recovering their value from the history of evaluated points.
+    :param configurations: list of configurations to evaluate.
+    :param hypermapper_mode: which HyperMapper mode is being used.
+    hypermapper_mode == "default"
+    :param param_space: a space object containing the search space.
+    :param beginning_of_time: timestamp of when the optimization started.
+    :param run_directory: directory where HyperMapper is running.
+    :param evolution_data_array: a dictionary containing all of the configurations that have been evaluated.
+    :param fast_addressing_of_data_array: a dictionary containing evaluated configurations and their index in
+    the evolution_data_array.
+    :param enable_feasible_predictor: whether to use constrained optimization.
+    :param evaluation_limit: the maximum number of function evaluations allowed for the evolutionary search.
+    :param black_box_function: the black_box_function being optimized in the evolutionary search.
+    :param number_of_cpus: an integer for the number of cpus to be used in parallel.
+    :return: configurations with evaluations for all points in configurations and the number of evaluated configurations
+    """
+    new_configurations = []
+    new_evaluations = {}
+    previous_evaluations = defaultdict(list)
+    number_of_new_evaluations = 0
+    t0 = datetime.datetime.now()
+    absolute_configuration_index = len(fast_addressing_of_data_array)
+
+    # Adds configutations to new if they have not been evaluated before
+    for configuration in configurations:
+        str_data = param_space.get_unique_hash_string_from_values(configuration)
+        if str_data in fast_addressing_of_data_array:
+            configuration_idx = fast_addressing_of_data_array[str_data]
+            for key in evolution_data_array:
+                previous_evaluations[key].append(evolution_data_array[key][configuration_idx])
+        else:
+            if absolute_configuration_index + number_of_new_evaluations < evaluation_limit:
+                new_configurations.append(configuration)
+                number_of_new_evaluations += 1
+
+    # Evaluates new configurations. If there is any
+    t1 = datetime.datetime.now()
+    if number_of_new_evaluations > 0:
+        new_evaluations = param_space.run_configurations(hypermapper_mode, new_configurations, beginning_of_time,
+                                                         black_box_function, run_directory)
```

Comment:
```suggestion
                                                         black_box_function, run_directory)
```
```suggestion
                                                         black_box_function, run_directory=run_directory)
```

Reply:
run_directory here must be a keyword parameter otherwise client-server mode does not work. Also, I imagine that Exhaustive mode is not implemented either, is that right?